node/lib/cmd/reset.js

```diff
@@ -0,0 +1,163 @@
+working directory for some of all of the tree.  This command can be used to
+(simultaneuosly or separately) unstage files, to undo changes to the working
+directory, and to change the commit pointed to by the current HEAD.  This
+command should not be confused with the 'revert' command.  The 'reset'
+command can be used to return a tree to a previous state in history; the
+'revert' command, on the other hand, creates new commits that undo previous
+changes.  After reseting the HEAD of the meta-repository, this command will
+apply the same change to all open sub-repositories, changing their HEADs to
+the commit indicated by the (new) HEAD in the meta-repository.  The index
+and working directory of meta- and sub-repositories may or may not be affected
+based on which mode (soft, mixed, or hard) is used; see the documentation
+for each mode for more information.
+`;
+
+/**
+ * Configure the specified `parser` for the `reset` command.
+ *
+ * @param {ArgumentParser} parser
+ */
+exports.configureParser = function (parser) {
+
+    parser.addArgument(["commitish"], {
+        type: "string",
+        help: "commit to reset the head of the current branch to",
+        defaultValue: null,
+        required: false,
+        nargs: "?",
+    });
+
+    parser.addArgument(["--soft"], {
+        required: false,
+        action: "storeConst",
+        constant: true,
+        help: `Do not change the working directory or index.`,
+    });
+
+    parser.addArgument(["--mixed"], {
+        required: false,
+        action: "storeConst",
+        constant: true,
+        help: `Unstage any changes, but to not alter the working dictory.  \
+This mode is the default.`,
+    });
+
+    parser.addArgument(["--hard"], {
+        required: false,
+        action: "storeConst",
+        constant: true,
+        help: `Discard all changes and reset both working directory and index \
+as specified by the selected commit.`
+    });
+};
+
+/**
+ * Execute the `reset` command according to the specified `args`.
+ *
+ * @async
+ * @param {Object}  args
+ * @param {String}  [args.commitish]
+ * @param {Boolean} args.mixed
+ */
+exports.executeableSubcommand = co.wrap(function *(args) {
+    const colors    = require("colors");
+
+    const GitUtil   = require("../util/git_util");
+    const Reset     = require("../util/reset");
+    const UserError = require("../util/user_error");
+    const Status    = require("../util/status");
+
+    const repo = yield GitUtil.getCurrentRepo();
+
+    // Figure out which commit to reset to.  If a commit was specified, look it
+    // up.  Otherwise, use HEAD.
+
+    let commit;
+    if (args.commitish) {
+
+        const commitish = args.commitish;
+        const annotated = yield GitUtil.resolveCommitish(repo, commitish);
+
+        if (null === annotated) {
+            throw new UserError(`Could not resolve ${colors.red(commitish)}.`);
+        }
+        commit = yield repo.getCommit(annotated.id());
```

Comment:
This method does throw (I wish it returned null/error instead), but I think we can assume that the fact that the `AnnotatedCommit` is valid (i.e., we resolve the comitish) implies the existence of a valid commit.